pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,238 @@
+
+import (
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+// LatencyStore is a store for latency information of connections between Nodes.
+type LatencyStore struct {
+	// Lock for the latency store
+	mutex sync.RWMutex
+
+	// Whether the agent is running in networkPolicyOnly mode
+	isNetworkPolicyOnly bool
+	// The map of Node IP to latency entry, it will be changed by latency monitor
+	nodeIPLatencyMap map[string]*NodeIPLatencyEntry
+	// The map of Node name to Node IP(s), it will be changed by Node watcher
+	// If the agent is running in networkPolicyOnly mode, the value will be the transport IP of the Node.
+	// Otherwise, the value will be the gateway IP of the Node
+	nodeTargetIPsMap map[string][]net.IP
+}
+
+// NodeIPLatencyEntry is the entry of the latency map.
+type NodeIPLatencyEntry struct {
+	// The timestamp of the last sent packet
+	LastSendTime time.Time
+	// The timestamp of the last received packet
+	LastRecvTime time.Time
+	// The last valid rtt of the connection
+	LastMeasuredRTT time.Duration
+}
+
+// NewLatencyStore creates a new LatencyStore.
+func NewLatencyStore(isNetworkPolicyOnly bool) *LatencyStore {
+	store := &LatencyStore{
+		nodeIPLatencyMap:    make(map[string]*NodeIPLatencyEntry),
+		nodeTargetIPsMap:    make(map[string][]net.IP),
+		isNetworkPolicyOnly: isNetworkPolicyOnly,
+	}
+
+	return store
+}
+
+// getNodeIPLatencyEntry returns the NodeIPLatencyEntry for the given Node IP
+// For now, it is only used for testing purposes.
+func (l *LatencyStore) getNodeIPLatencyEntry(nodeIP string) (*NodeIPLatencyEntry, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	entry, ok := l.nodeIPLatencyMap[nodeIP]
+
+	return entry, ok
+}
+
+// DeleteNodeIPLatencyEntry deletes the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) DeleteNodeIPLatencyEntry(nodeIP string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.nodeIPLatencyMap, nodeIP)
+}
+
+// SetNodeIPLatencyEntry sets the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) SetNodeIPLatencyEntry(nodeIP string, mutator func(entry *NodeIPLatencyEntry)) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	entry, ok := l.nodeIPLatencyMap[nodeIP]
+	if !ok {
+		entry = &NodeIPLatencyEntry{}
+		l.nodeIPLatencyMap[nodeIP] = entry
+	}
+
+	mutator(entry)
+}
+
+// addNode adds a Node to the latency store
+func (l *LatencyStore) addNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (l *LatencyStore) deleteNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	nodeIPs, ok := l.nodeTargetIPsMap[node.Name]
+	if !ok {
+		klog.ErrorS(nil, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	for _, ip := range nodeIPs {
+		delete(l.nodeIPLatencyMap, ip.String())
+	}
```

Comment:
That's a good point. What solution do you have in mind?
1) we could add the Node name to the `NodeIPLatencyEntry` value in `l.nodeIPLatencyMap` and skip deleting it if the names don't match
2) we could follow the same approach as the NodeRouteController: add a workqueue and retry the ADD event until the previous DELETE goes through

Personally I think that because of the nature of this controller, deleting an entry in `l. nodeIPLatencyMap` does not matter too much. We will add the entry "back" the next time a probe is sent. It doesn't really impact correctness the way correctness is impacted for the NodeRouteController. *So maybe we could add a comment about this issue and leave it as future work?*

Reply:
If `nodeIPLatencyMap` will be auto-corrected, I agree we don't need to worry about the case. On the other hand, should we defer adding/deleting items to `nodeIPLatencyMap` to `monitorLoop` and only update `nodeTargetIPsMap` in event handlers? I think it could cover this case and https://github.com/antrea-io/antrea/pull/6120#discussion_r1608072274.

Reply:
Isn't item deletion complicated to do in `monitorLoop`?

Reply:
I thought it only needs to remove keys that in `nodeIPLatencyMap` but not in `nodeIPs` which is calculated from `m.latencyStore.ListNodeIPs()`, no strong opinion though.

However, note that the removed IP may be added back to `nodeIPLatencyMap` if it receives a response after the IP is deleted, so either it should only update existing entries in `recvPing` or it should do IP cleanup in `monitorLoop` periodically.

Reply:
Deleting stale IPs after calling `ListNodeIPs` sounds good to me for now.
Could you implement this @IRONICBo

Reply:
That's great, I will update it.